File: benchmarks/tsb/bench_api_types.ts
/**
 * Benchmark: api.types predicates — isScalar, isListLike, isNumericDtype, etc.
 * Runs 100,000 checks across value-level and dtype-level predicates.
 */
import { api } from "../../src/index.js";

const WARMUP = 3;
const ITERATIONS = 30;

const values = [42, "hello", true, null, [1, 2, 3], { a: 1 }, new Date(), 3.14, BigInt(99)];
const dtypes = ["int64", "float64", "bool", "object", "string", "datetime64[ns]", "category", "interval"];

function runChecks(): void {
  for (const v of values) {
    api.types.isScalar(v);
    api.types.isListLike(v);
    api.types.isArrayLike(v);
    api.types.isDictLike(v);
    api.types.isNumber(v);
    api.types.isBool(v);
    api.types.isMissing(v);
  }
  for (const d of dtypes) {
    api.types.isNumericDtype(d);
    api.types.isIntegerDtype(d);
    api.types.isFloatDtype(d);
    api.types.isBoolDtype(d);
    api.types.isDatetimeDtype(d);
    api.types.isCategoricalDtype(d);
    api.types.isObjectDtype(d);
  }
}

for (let i = 0; i < WARMUP; i++) {
  runChecks();
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  runChecks();
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "api_types",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_api_types.py
"""Benchmark: pandas.api.types predicates — is_scalar, is_list_like, is_numeric_dtype, etc."""
import json
import time
import pandas as pd
import numpy as np

WARMUP = 3
ITERATIONS = 30

values = [42, "hello", True, None, [1, 2, 3], {"a": 1}, pd.Timestamp("2020-01-01"), 3.14]
dtypes = [
    np.dtype("int64"),
    np.dtype("float64"),
    np.dtype("bool"),
    np.dtype("O"),
    pd.StringDtype(),
    np.dtype("datetime64[ns]"),
    pd.CategoricalDtype(),
    pd.IntervalDtype(),
]


def run_checks() -> None:
    for v in values:
        pd.api.types.is_scalar(v)
        pd.api.types.is_list_like(v)
        pd.api.types.is_dict_like(v)
        pd.api.types.is_number(v)
        pd.api.types.is_bool(v)
    for d in dtypes:
        pd.api.types.is_numeric_dtype(d)
        pd.api.types.is_integer_dtype(d)
        pd.api.types.is_float_dtype(d)
        pd.api.types.is_bool_dtype(d)
        pd.api.types.is_datetime64_any_dtype(d)
        pd.api.types.is_categorical_dtype(d)
        pd.api.types.is_object_dtype(d)


for _ in range(WARMUP):
    run_checks()

start = time.perf_counter()
for _ in range(ITERATIONS):
    run_checks()
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "api_types",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
